orttraining/orttraining/python/orttraining_python_module.cc

```diff
@@ -357,18 +357,23 @@ PYBIND11_MODULE(onnxruntime_pybind11_state, m) {
 
   m.def("get_version_string", []() -> std::string { return ORT_VERSION; });
 
-  m.def("clear_training_ep_instances", []() -> void {
-    ort_training_env->ClearExecutionProviderInstances();
-  },
-        "Clean the execution provider instances used in ort training module.");
+  m.def(
+      "clear_training_ep_instances", []() -> void {
+        ort_training_env->ClearExecutionProviderInstances();
+      },
+      "Clean the execution provider instances used in ort training module.");
 
-  // clean the ort training environment when python interpreter exit
-  // otherwise the global var will be de-constrcut after user main.
-  // the order of ort training environment deconstruction and cudart
-  // deconstruction is not stable, which will lead to crash.
   auto atexit = py::module_::import("atexit");
   atexit.attr("register")(py::cpp_function([]() {
-    ort_training_env = nullptr;
+    // Clean the execution providers map stored in the ort training environment
+    // when python interpreter exit otherwise the global var's execution provider map
+    // will be de-constrcut after user main.
+    // Otherwise, the order of ort training environment deconstruction and cudart
+    // deconstruction is not stable, which will lead to crash:
+    // CUDA failure 4 : driver shutting down
+    // Make sure not to destruct the environment itself yet. Otherwise, the environment may
```

Comment:
It's not obvious to me why the sessions would outlive the environment. In a typical Python program I create a session, do some work with the session and the Python program exits. On exit I would expect the session to get destructed first. Please include an explanation since a fair amount of investigation has gone into this and it'll be helpful to educate new folks.

Reply:
I do not have a good answer for this yet. So far, all I know is that the environment on occasion gets destroyed before the session. 
Let me find the answer to why that might happen.